---
title: "metaWRAP workflow"
author: "xyz"
date: "2021/5/26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# QC

[fastp](https://github.com/OpenGene/fastp)

qualified_quality_phred
  -q 20 
unqualified_percent_limit
  -u 20 
average_quality
  -e 30 

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/ammoxidationSIPheavyAOA
#PBS -j oe
#PBS -l nodes=1:ppn=16,walltime=24:24:00,mem=30gb
#PBS -N fastp
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd ~/huangxueru/氨氧化-SIP-重层-AOA标记上/Rawdata
fastp -i L1HFC080438_CC_S94_L002_R1_001.fastq.gz \
  -I L1HFC080438_CC_S94_L002_R2_001.fastq.gz \
  -o ~/ammoxidationSIPheavyAOA/S94.R1.fq.gz \
  -O ~/ammoxidationSIPheavyAOA/S94.R2.fq.gz \
  --thread 16\
  -q 20 \
  -u 20 \
  -e 30 \
  -h ~/ammoxidationSIPheavyAOA/S94.html
wait
fastp -i L1HFC080439_CM_S95_L002_R1_001.fastq.gz \
  -I L1HFC080439_CM_S95_L002_R2_001.fastq.gz \
  -o ~/ammoxidationSIPheavyAOA/S95.R1.fq.gz \
  -O ~/ammoxidationSIPheavyAOA/S95.R2.fq.gz \
  --thread 16 \
  -q 20 \
  -u 20 \
  -e 30 \
  -h ~/ammoxidationSIPheavyAOA/S95.html
wait
fastp -i L1HFC080440_CS_S96_L002_R1_001.fastq.gz \
  -I L1HFC080440_CS_S96_L002_R2_001.fastq.gz \
  -o ~/ammoxidationSIPheavyAOA/S96.R1.fq.gz \
  -O ~/ammoxidationSIPheavyAOA/S96.R2.fq.gz \
  --thread 16 \
  -q 20 \
  -u 20 \
  -e 30 \
  -h ~/ammoxidationSIPheavyAOA/S96.html

vim fastpJob.sh 
qsub fastpJob.sh
# get job status
qstat -f 699077
# kill job
qdel 699077
```

# assemble by megahit

[megahit](https://github.com/voutcn/megahit)

meta '--min-count 2 --k-list 21,41,61,81,99' 
(generic metagenomes, default)

meta-sensitive '--min-count 2 --k-list 21,31,41,51,61,71,81,91,99' 
(more sensitive but slower)

meta-large '--min-count 2 --k-list 27,37,47,57,67,77,87' 
(large & complex metagenomes, like soil)

bulk '--min-count 3 --k-list 31,51,71,91,99 --no-mercy'  
(experimental, standard bulk sequencing with >= 30x depth)

single-cell '--min-count 3 --k-list 21,33,55,77,99,121 --merge_level 20,0.96' 
(experimental, single cell data)

138109 contigs, total 100213737 bp, min 200 bp, 
max 277744 bp, avg 725 bp, N50 1326 bp

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/ammoxidationSIPheavyAOA
#PBS -j oe
#PBS -l nodes=1:ppn=22,walltime=999:00:00,mem=120gb
#PBS -N megahit
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd ~/ammoxidationSIPheavyAOA/
megahit -1 S94.R1.fq.gz,S95.R1.fq.gz,S96.R1.fq.gz \
  -2 S94.R2.fq.gz,S95.R2.fq.gz,S96.R2.fq.gz \
  -o megahitOut \
  --min-count 2 \
  --k-list 27,37,47,57,67,77,87 \
  -t 22
  
vim megahitJob.sh 
qsub megahitJob.sh
# get job status
qstat -f 699088
# kill job
qdel 699088
# see the work load
pestat
```

# binning by metaWRAP

[metaWRAP](https://github.com/bxlab/metaWRAP)

[metaWRAP tutorial](https://github.com/bxlab/metaWRAP/blob/master/Usage_tutorial.md)

[kraken2 Manual](https://github.com/DerrickWood/kraken2/wiki/Manual)

[metaWRAP Chinese tutorial](https://mp.weixin.qq.com/s/Ecn4DOrhfUhz1HynbgQtnw)

[invalid compressed data--format violated](https://github.com/DerrickWood/kraken2/issues/412)

```{bash}
# checkM database
mkdir CHECKM_DB
cd CHECKM_DB
wget https://data.ace.uq.edu.au/public/CheckM_databases/checkm_data_2015_01_16.tar.gz
tar -xvf *.tar.gz
rm *.gz
cd ..
checkm data setRoot CHECKM_DB

# unzip fastaq
cd /home/xiongyi/ammoxidationSIPheavyAOA/

#!/bin/sh -login
#PBS -o /home/xiongyi/ammoxidationSIPheavyAOA/cleanFastaq/
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N unzip-fastaq
#PBS -q cpu 
cd /home/xiongyi/ammoxidationSIPheavyAOA/
gunzip -c S94.R1.fq.gz >cleanFastaq/S94_1.fastq &
gunzip -c S94.R2.fq.gz >cleanFastaq/S94_2.fastq 

gunzip -c S95.R1.fq.gz >cleanFastaq/S95_1.fastq &
gunzip -c S95.R2.fq.gz >cleanFastaq/S95_2.fastq 

gunzip -c S96.R1.fq.gz >cleanFastaq/S96_1.fastq &
gunzip -c S96.R2.fq.gz >cleanFastaq/S96_2.fastq

vim unzip-fastaq.sh
qsub unzip-fastaq.sh

qstat -f 699118

#!/bin/sh -login
#PBS -o /home/xiongyi/ammoxidationSIPheavyAOA/INITIAL_BINNING
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N metawrap-binning
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/ammoxidationSIPheavyAOA/
metawrap binning -o INITIAL_BINNING -t 24 \
  -a megahitOut/final.contigs.fa \
  --metabat2 --maxbin2 \
  --concoct cleanFastaq/*.fastq \
  --run-checkm
  
vim metawrap-binning.sh 
qsub metawrap-binning.sh
# get job status
qstat -f 699112
# kill job
qdel 699112
# see the work load
pestat
```

# species annotation by kraken2

```{bash install}
conda create -y -n metawrap-env python=2.7
conda activate metawrap-env
conda config --add channels defaults
conda config --add channels conda-forge
conda config --add channels bioconda
conda config --add channels ursky
conda install --only-deps -c ursky metawrap-mg
conda install -c ursky metawrap-mg
conda install -c bioconda kraken2
# fix dustmasker: error while loading shared libraries
conda install -c bioconda blast openssl=1.0
# update to newest version
conda install -y -c ursky metawrap-mg=1.3.2

mkdir /home/xiongyi/database/KRAKEN_DB
cd /home/xiongyi/database/KRAKEN_DB
kraken2-build --standard --threads 12 --db MY_KRAKEN2_DB

# fix problem of abortive download
# only download plasmid db and build again
# try 3 times, cannot work
kraken2-build --download-library plasmid --db MY_KRAKEN2_DB --use-ftp

kraken2-build --download-library fungi --db MY_KRAKEN2_DB

#!/bin/sh -login
#PBS -o /home/xiongyi/database/KRAKEN_DB
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N kraken2-build
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/database/KRAKEN_DB
kraken2-build --build --db MY_KRAKEN2_DB --threads 24
  
vim kraken2-build.sh 
qsub kraken2-build.sh
# get job status
qstat -f 699112
# kill job
qdel 699112
# see the work load
pestat

#!/bin/sh -login
#PBS -o /home/xiongyi/ammoxidationSIPheavyAOA/KRAKENreads
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N kraken-reads-run
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/ammoxidationSIPheavyAOA/
metawrap kraken2 -o KRAKENreads -t 24 cleanFastaq/*fastq megahitOut/final.contigs.fa

vim kraken-reads-run.sh 
qsub kraken-reads-run.sh
qstat -f 699133
qdel 699133
```

# refine bin

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/ammoxidationSIPheavyAOA/BIN_REFINEMENT
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N bin_refinement
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/ammoxidationSIPheavyAOA/
metawrap bin_refinement -o BIN_REFINEMENT \
  -t 24 \
  -A INITIAL_BINNING/metabat2_bins/ \
  -B INITIAL_BINNING/maxbin2_bins/ \
  -c 50 -x 10

vim bin_refinement.sh 
qsub bin_refinement.sh
qstat -f 699137
qdel 699137
```

# Visualize bins

[install aspera](https://www.jianshu.com/p/240b7d751cf1)

```{bash}
ftp
open ftp.ncbi.nlm.nih.gov
# user
anonymous
cd /blast/db/
passive
mls nt.*.tar.gz download.list.txt
cd /blast/db/v4/
mls nt_v4.*.tar.gz downloadV4.list.txt
bye

nohup cat NCBI.nt.download.list.txt | \
  xargs -n 1 -P 1 \
  bash -c '~/.aspera/connect/bin/ascp -v -k 1 -T -l 1000m -i ~/asperaweb_id_dsa.openssh anonftp@ftp.ncbi.nlm.nih.gov:/blast/db/$0 ./' \
  >> downloadlog.txt 2>&1 &
  
nohup cat downloadV4.list.txt | \
  xargs -n 1 -P 1 \
  bash -c '~/.aspera/connect/bin/ascp -v -k 1 -T -l 1000m -i ~/asperaweb_id_dsa.openssh anonftp@ftp.ncbi.nlm.nih.gov:/blast/db/v4/$0 ./' \
  >> downloadlog.txt 2>&1 &
  
~/.aspera/connect/bin/ascp -v -k 1 -T -l 1000m \
  -i ~/asperaweb_id_dsa.openssh \
  anonftp@ftp.ncbi.nlm.nih.gov:/pub/taxonomy/taxdump.tar.gz ./
tar -xvf taxdump.tar.gz

#!/bin/sh -login
#PBS -o /home/xiongyi/ammoxidationSIPheavyAOA/BLOBOLOGY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BLOBOLOGY
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/ammoxidationSIPheavyAOA/
metawrap blobology -a megahitOut/final.contigs.fa \
  -t 24 \
  -o BLOBOLOGY \
  --bins BIN_REFINEMENT/metawrap_50_10_bins \
  cleanFastaq/*fastq

vim blobology.sh 
qsub blobology.sh
qstat -f 699137
qdel 699137
```

