---
title: "Methane Oxidation"
author: "xyz"
date: "2021/6/9"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# unzip data

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/cleanFastaq
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N unzip-fastaq
#PBS -q cpu 
cd /home/xiongyi/MethaneOxidation/cleanFastaq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S46/S46_350.fq1.gz >S46_1.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S46/S46_350.fq2.gz >S46_2.fastq 
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S50/S50_350.fq1.gz >S50_1.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S50/S50_350.fq2.gz >S50_2.fastq 
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S54/S54_350.fq1.gz >S54_1.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S54/S54_350.fq2.gz >S54_2.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S67/S67_350.fq1.gz >S67_1.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S67/S67_350.fq2.gz >S67_2.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S71/S71_350.fq1.gz >S71_1.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S71/S71_350.fq2.gz >S71_2.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S72/S72_350.fq1.gz >S72_1.fastq
gunzip -c /home/xiongyi/huangxueru/甲烷氧化/CleanData/S72/S72_350.fq2.gz >S72_2.fastq

vim unzip-fastaq.sh
qsub unzip-fastaq.sh
```

# assemble

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/ASSEMBLY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N merge
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/MethaneOxidation/
cat cleanFastaq/*_1.fastq > cleanFastaq/All1.fastq
cat cleanFastaq/*_2.fastq > cleanFastaq/All2.fastq

vim merge.sh

# use fat node

#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/ASSEMBLY
#PBS -j oe
#PBS -l nodes=node31:ppn=24,walltime=999:00:00,mem=250gb
#PBS -N assembly
#PBS -q batch
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/MethaneOxidation/
metawrap assembly \
  -1 cleanFastaq/All1.fastq \
  -2 cleanFastaq/All2.fastq \
  -m 249 \
  -t 24 \
  --megahit \
  -o ASSEMBLY
  
vim assembly.sh
qsub assembly.sh

#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/ASSEMBLY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N assembly
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/MethaneOxidation/
metawrap assembly \
  -1 cleanFastaq/All1.fastq \
  -2 cleanFastaq/All2.fastq \
  -m 120 \
  -t 24 \
  --megahit \
  -o ASSEMBLY
  
vim assembly.sh
qsub assembly.sh

# use megahit

#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/ASSEMBLY
#PBS -j oe
#PBS -l nodes=node31:ppn=24,walltime=999:00:00,mem=250gb
#PBS -N megahit
#PBS -q batch
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd /home/xiongyi/MethaneOxidation/
megahit -1 cleanFastaq/All1.fastq \
  -2 cleanFastaq/All2.fastq \
  -o ASSEMBLY \
  --min-count 2 \
  -m 249 \
  --k-list 27,37,47,57,67,77,87 \
  -t 24
vim megahit.sh
```

# bining

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/INITIAL_BINNING
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N metawrap-binning
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/MethaneOxidation/
metawrap binning -o INITIAL_BINNING -t 24 \
  -a ASSEMBLY/final.contigs.fa \
  --metabat2 --maxbin2 \
  --concoct cleanFastaq/S*.fastq \
  --run-checkm
  
vim metawrap-binning.sh 
qsub metawrap-binning.sh

# no maxbin maxbin is too slow

#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/INITIAL_BINNING_no_maxbin
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N metawrap-binning
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/MethaneOxidation/
metawrap binning -o INITIAL_BINNING_no_maxbin -t 24 \
  -a ASSEMBLY/final.contigs.fa \
  --metabat2 \
  --concoct cleanFastaq/S*.fastq \
  --run-checkm
  
vim metawrap-binning_no_maxbin.sh 
qsub metawrap-binning_no_maxbin.sh
```

# species annotation

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/KRAKENreads
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N kraken-reads-run
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap-env
cd /home/xiongyi/MethaneOxidation/
metawrap kraken2 -o KRAKENreads -t 24 cleanFastaq/S*.fastq ASSEMBLY/final.contigs.fa

vim kraken-reads-run.sh 
qsub kraken-reads-run.sh
```

# refine bin

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/BIN_REFINEMENT
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N bin_refinement
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/MethaneOxidation/
metawrap bin_refinement -o BIN_REFINEMENT \
  -t 24 \
  -A INITIAL_BINNING_no_maxbin/metabat2_bins/ \
  -B INITIAL_BINNING_no_maxbin/concoct_bins/ \
  -c 50 -x 10

vim bin_refinement.sh 
qsub bin_refinement.sh
```

# Visualize bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/BLOBOLOGY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BLOBOLOGY
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/MethaneOxidation/
metawrap blobology -a ASSEMBLY/final.contigs.fa \
  -t 24 \
  -o BLOBOLOGY \
  --bins BIN_REFINEMENT/metawrap_50_10_bins \
  cleanFastaq/*_*.fastq

vim blobology.sh 
qsub blobology.sh
```

# get abundance of bin

error
128g ram is not enough

processed 1,726,000,000 
positions/home/xiongyi/miniconda3/envs/metawrap/bin/metawrap-modules/quant_bins.sh: 
line 128:  3623 Killed
salmon index -p $threads -t $assembly -i ${out}/assembly_index

use less thread to reduce ram suage

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/QUANT_BINS
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=125gb
#PBS -N QUANT_BINS
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/MethaneOxidation/
metawrap quant_bins -b BIN_REFINEMENT/metawrap_50_10_bins \
  -o QUANT_BINS \
  -t 12 \
  -a ASSEMBLY/final.contigs.fa cleanFastaq/*_*.fastq
  
vim quant.sh 
```

# Re-assemble bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/BIN_REASSEMBLY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BIN_REASSEMBLY
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/MethaneOxidation/
metawrap reassemble_bins \
  -1 cleanFastaq/All1.fastq \
  -2 cleanFastaq/All2.fastq \
  -t 24 \
  -m 120 \
  -c 50 \
  -x 10 \
  -o BIN_REASSEMBLY \
  -b BIN_REFINEMENT/metawrap_50_10_bins
  
vim reassemble_bins.sh 
```

# species annotation of bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/BIN_CLASSIFICATION
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BIN_CLASSIFICATION
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/MethaneOxidation/
metawrap classify_bins -b BIN_REASSEMBLY/reassembled_bins \
  -o BIN_CLASSIFICATION \
  -t 24
  
vim classfication_bins.sh 
```

# function annotation of bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/MethaneOxidation/FUNCT_ANNOT
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BIN_CLASSIFICATION
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/MethaneOxidation/
metaWRAP annotate_bins -o FUNCT_ANNOT -t 24 -b BIN_REASSEMBLY/reassembled_bins/

vim annotate_bins.sh
```

