---
title: "cai"
author: "xyz"
date: "2021/8/27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# unzip

```{bash}
cd /home/xiongyi/sunhongyang/MbPL2020111531
mkdir cleanFastaq KRAKENreads \
  INITIAL_BINNING BIN_REFINEMENT \
  BIN_REASSEMBLY BLOBOLOGY QUANT_BINS \
  BIN_CLASSIFICATION FUNCT_ANNOT QUANT_CONTIG

#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/cleanFastaq
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N unzip-fastaq
#PBS -q cpu 
cd /home/xiongyi/sunhongyang/MbPL2020111531/cleanFastaq
gunzip -c /home/xiongyi/sunhongyang/MbPL2020111531/RAW/1_T_DNA_clean_R1.fq.gz > T1_1.fastq
gunzip -c /home/xiongyi/sunhongyang/MbPL2020111531/RAW/1_T_DNA_clean_R2.fq.gz > T1_2.fastq
gunzip -c /home/xiongyi/sunhongyang/MbPL2020111531/RAW/4_T_DNA_clean_R1.fq.gz > T4_1.fastq
gunzip -c /home/xiongyi/sunhongyang/MbPL2020111531/RAW/4_T_DNA_clean_R2.fq.gz > T4_2.fastq
gunzip -c /home/xiongyi/sunhongyang/MbPL2020111531/RAW/7_T_DNA_clean_R1.fq.gz > T7_1.fastq
gunzip -c /home/xiongyi/sunhongyang/MbPL2020111531/RAW/7_T_DNA_clean_R2.fq.gz > T7_2.fastq

vim unzip-fastaq.sh
qsub unzip-fastaq.sh
```

# assemble

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/cleanFastaq
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N merge
#PBS -q cpu
cd /home/xiongyi/sunhongyang/MbPL2020111531/
cat cleanFastaq/*_1.fastq > cleanFastaq/All1.fastq
cat cleanFastaq/*_2.fastq > cleanFastaq/All2.fastq

vim merge.sh

#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=124gb
#PBS -N megahit
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd /home/xiongyi/sunhongyang/MbPL2020111531/
megahit -1 cleanFastaq/All1.fastq \
  -2 cleanFastaq/All2.fastq \
  -o ASSEMBLY \
  --min-count 2 \
  -m 123 \
  --k-list 27,37,47,57,67,77,87 \
  -t 24
  
cd /home/xiongyi/sunhongyang/MbPL2020111531/
vim megahit.sh
```

# QUANT_CONTIG

[get file name in bash](https://stackoverflow.com/questions/10981439/reading-filenames-into-an-array)

[get unique value of bash array](https://stackoverflow.com/questions/13648410/how-can-i-get-unique-values-from-an-array-in-bash)

[get substring of value](https://stackoverflow.com/questions/22404139/how-to-get-substring-from-filename-in-unix-shellscripting)
a=25_good_dfdfdf
echo ${a#*_}   # remove prefix ending in "_"
echo ${a%_*}   # remove suffix starting with "_"

```{bash}
conda create -c bioconda -n salmon salmon

#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/QUANT_CONTIG
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N QUANT_CONTIG
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate salmon
cd /home/xiongyi/sunhongyang/MbPL2020111531/
salmon index -t ASSEMBLY/final.contigs.fa \
  -i QUANT_CONTIG/contig_index \
  --threads 24
# get fastq name
files=(cleanFastaq/*_*.fastq)
# get sample name
files=($(echo "${files[@]}" | tr ' ' '\n' | cut -d '_' -f1 | tr '\n' ' '))
# unique sample name
files=($(echo "${files[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))
for file in ${files[@]}; do 
    salmon quant -i QUANT_CONTIG/contig_index \
    -l A \
    -1 ${file}_1.fastq \
    -2 ${file}_2.fastq \
    --validateMappings \
    -o QUANT_CONTIG/${file#*/} \
    --threads 24
done

vim QUANT_CONTIG.sh
qsub QUANT_CONTIG.sh
```

# bining

no maxbin because maxbin is too slow

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/INITIAL_BINNING
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N metawrap-binning
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/sunhongyang/MbPL2020111531/
metawrap binning -o INITIAL_BINNING -t 24 \
  -a ASSEMBLY/final.contigs.fa \
  --metabat2 \
  --concoct cleanFastaq/*_*.fastq \
  --run-checkm
  
vim binning.sh 
qsub binning.sh
```

# species annotation

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/KRAKENreads
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N kraken-reads-run
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/sunhongyang/MbPL2020111531/
metawrap kraken2 -o KRAKENreads -t 24 cleanFastaq/*_*.fastq ASSEMBLY/final.contigs.fa

vim kraken-reads-run.sh 
qsub kraken-reads-run.sh
```

# refine bin

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/BIN_REFINEMENT
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N bin_refinement
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/sunhongyang/MbPL2020111531/
metawrap bin_refinement -o BIN_REFINEMENT \
  -t 24 \
  -A INITIAL_BINNING/metabat2_bins \
  -B INITIAL_BINNING/concoct_bins \
  -c 50 -x 10

vim bin_refinement.sh 
qsub bin_refinement.sh
```

# Visualize bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/BLOBOLOGY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BLOBOLOGY
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/sunhongyang/MbPL2020111531/
metawrap blobology -a ASSEMBLY/final.contigs.fa \
  -t 24 \
  -o BLOBOLOGY \
  --bins BIN_REFINEMENT/metawrap_50_10_bins \
  cleanFastaq/*_*.fastq

vim blobology.sh 
qsub blobology.sh
```

# species annotation of bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/BIN_CLASSIFICATION
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BIN_CLASSIFICATION
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/sunhongyang/MbPL2020111531/
metawrap classify_bins -b BIN_REFINEMENT/metawrap_50_10_bins \
  -o BIN_CLASSIFICATION \
  -t 24
  
vim classfication_bins.sh 
```

# function annotation of bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/sunhongyang/MbPL2020111531/FUNCT_ANNOT
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BIN_FUNCT_ANNOT
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/sunhongyang/MbPL2020111531/
metaWRAP annotate_bins -o FUNCT_ANNOT -t 24 -b BIN_REFINEMENT/metawrap_50_10_bins

vim annotate_bins.sh
```



