---
title: "YY"
author: "xyz"
date: "2021/7/4"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# unzip

```{r}
# ls ~/YY/A*/A*.gz -l > cleanFastaq/fileList.txt
library(stringr)
df<-read.table("../data/YYfileList.txt")
newName<-str_split(df$V9,"/",simplify = T)[,5]
newName<-paste0(newName,rep(c("_1","_2"),18),".fastq")
cmd<-paste0("gunzip -c ",df$V9," > ",newName)
writeLines(cmd,"../data/YYcmd.txt")
```

```{bash}
cd /home/xiongyi/YYDS/
mkdir cleanFastaq KRAKENreads \
  INITIAL_BINNING BIN_REFINEMENT \
  BIN_REASSEMBLY BLOBOLOGY QUANT_BINS \
  BIN_CLASSIFICATION FUNCT_ANNOT

#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/cleanFastaq
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N unzip-fastaq
#PBS -q cpu 
cd /home/xiongyi/YYDS/cleanFastaq
gunzip -c /home/xiongyi/YY/A13/A13_350.fq1.gz > A13_1.fastq
gunzip -c /home/xiongyi/YY/A13/A13_350.fq2.gz > A13_2.fastq
gunzip -c /home/xiongyi/YY/A14/A14_350.fq1.gz > A14_1.fastq
gunzip -c /home/xiongyi/YY/A14/A14_350.fq2.gz > A14_2.fastq
gunzip -c /home/xiongyi/YY/A19/A19_350.fq1.gz > A19_1.fastq
gunzip -c /home/xiongyi/YY/A19/A19_350.fq2.gz > A19_2.fastq
gunzip -c /home/xiongyi/YY/A1/A1_350.fq1.gz > A1_1.fastq
gunzip -c /home/xiongyi/YY/A1/A1_350.fq2.gz > A1_2.fastq
gunzip -c /home/xiongyi/YY/A20/A20_350.fq1.gz > A20_1.fastq
gunzip -c /home/xiongyi/YY/A20/A20_350.fq2.gz > A20_2.fastq
gunzip -c /home/xiongyi/YY/A25/A25_350.fq1.gz > A25_1.fastq
gunzip -c /home/xiongyi/YY/A25/A25_350.fq2.gz > A25_2.fastq
gunzip -c /home/xiongyi/YY/A26/A26_350.fq1.gz > A26_1.fastq
gunzip -c /home/xiongyi/YY/A26/A26_350.fq2.gz > A26_2.fastq
gunzip -c /home/xiongyi/YY/A2/A2_350.fq1.gz > A2_1.fastq
gunzip -c /home/xiongyi/YY/A2/A2_350.fq2.gz > A2_2.fastq
gunzip -c /home/xiongyi/YY/A31/A31_350.fq1.gz > A31_1.fastq
gunzip -c /home/xiongyi/YY/A31/A31_350.fq2.gz > A31_2.fastq
gunzip -c /home/xiongyi/YY/A32/A32_350.fq1.gz > A32_1.fastq
gunzip -c /home/xiongyi/YY/A32/A32_350.fq2.gz > A32_2.fastq
gunzip -c /home/xiongyi/YY/A37/A37_350.fq1.gz > A37_1.fastq
gunzip -c /home/xiongyi/YY/A37/A37_350.fq2.gz > A37_2.fastq
gunzip -c /home/xiongyi/YY/A38/A38_350.fq1.gz > A38_1.fastq
gunzip -c /home/xiongyi/YY/A38/A38_350.fq2.gz > A38_2.fastq
gunzip -c /home/xiongyi/YY/A43/A43_350.fq1.gz > A43_1.fastq
gunzip -c /home/xiongyi/YY/A43/A43_350.fq2.gz > A43_2.fastq
gunzip -c /home/xiongyi/YY/A44/A44_350.fq1.gz > A44_1.fastq
gunzip -c /home/xiongyi/YY/A44/A44_350.fq2.gz > A44_2.fastq
gunzip -c /home/xiongyi/YY/A49/A49_350.fq1.gz > A49_1.fastq
gunzip -c /home/xiongyi/YY/A49/A49_350.fq2.gz > A49_2.fastq
gunzip -c /home/xiongyi/YY/A50/A50_350.fq1.gz > A50_1.fastq
gunzip -c /home/xiongyi/YY/A50/A50_350.fq2.gz > A50_2.fastq
gunzip -c /home/xiongyi/YY/A7/A7_350.fq1.gz > A7_1.fastq
gunzip -c /home/xiongyi/YY/A7/A7_350.fq2.gz > A7_2.fastq
gunzip -c /home/xiongyi/YY/A8/A8_350.fq1.gz > A8_1.fastq
gunzip -c /home/xiongyi/YY/A8/A8_350.fq2.gz > A8_2.fastq

vim unzip-fastaq.sh
qsub unzip-fastaq.sh
```

# assemble

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/ASSEMBLY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N merge
#PBS -q cpu
cd /home/xiongyi/YYDS/
cat cleanFastaq/*_1.fastq > cleanFastaq/All1.fastq
cat cleanFastaq/*_2.fastq > cleanFastaq/All2.fastq

vim merge.sh

#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/
#PBS -j oe
#PBS -l nodes=node31:ppn=24,walltime=999:00:00,mem=250gb
#PBS -N megahit
#PBS -q batch
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd /home/xiongyi/YYDS/
megahit -1 cleanFastaq/All1.fastq \
  -2 cleanFastaq/All2.fastq \
  -o ASSEMBLY \
  --min-count 2 \
  -m 249 \
  --k-list 27,37,47,57,67,77,87 \
  -t 24

cd /home/xiongyi/YYDS/
vim megahit.sh
```

# quantify contig

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/QUANT_CONTIG
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N QUANT_CONTIG
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate salmon
cd /home/xiongyi/YYDS/
salmon index -t ASSEMBLY/final.contigs.fa \
  -i QUANT_CONTIG/contig_index \
  --threads 24
files=(cleanFastaq/*_*.fastq)
files=($(echo "${files[@]}" | tr ' ' '\n' | cut -d '_' -f1 | tr '\n' ' '))
files=($(echo "${files[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))
for file in ${files[@]}; do 
    salmon quant -i QUANT_CONTIG/contig_index \
    -l A \
    -1 ${file}_1.fastq \
    -2 ${file}_2.fastq \
    --validateMappings \
    -o QUANT_CONTIG/${file#*/} \
    --threads 24
done

vim QUANT_CONTIG.sh
qsub QUANT_CONTIG.sh
```

# bining

no maxbin because maxbin is too slow

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/INITIAL_BINNING
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N metawrap-binning
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/YYDS/
metawrap binning -o INITIAL_BINNING -t 24 \
  -a ASSEMBLY/final.contigs.fa \
  --metabat2 \
  --concoct cleanFastaq/*_*.fastq \
  --run-checkm
  
vim binning.sh 
qsub binning.sh
```

# species annotation

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/KRAKENreads
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N kraken-reads-run
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/YYDS/
metawrap kraken2 -o KRAKENreads -t 24 cleanFastaq/*_*.fastq ASSEMBLY/final.contigs.fa

vim kraken-reads-run.sh 
qsub kraken-reads-run.sh
```

# refine bin

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/BIN_REFINEMENT
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N bin_refinement
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/YYDS/
metawrap bin_refinement -o BIN_REFINEMENT \
  -t 24 \
  -A INITIAL_BINNING/metabat2_bins/ \
  -B INITIAL_BINNING/concoct_bins/ \
  -c 50 -x 10

vim bin_refinement.sh 
qsub bin_refinement.sh
```

# Visualize bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/BLOBOLOGY
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BLOBOLOGY
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/YYDS/
metawrap blobology -a ASSEMBLY/final.contigs.fa \
  -t 24 \
  -o BLOBOLOGY \
  --bins BIN_REFINEMENT/metawrap_50_10_bins \
  cleanFastaq/*_*.fastq

vim blobology.sh 
qsub blobology.sh
```

# species annotation of bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/BIN_CLASSIFICATION
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BIN_CLASSIFICATION
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/YYDS/
metawrap classify_bins -b BIN_REFINEMENT/metawrap_50_10_bins \
  -o BIN_CLASSIFICATION \
  -t 24
  
vim classfication_bins.sh 
```

# function annotation of bins

```{bash}
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/FUNCT_ANNOT
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N BIN_FUNCT_ANNOT
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate metawrap
cd /home/xiongyi/YYDS/
metaWRAP annotate_bins -o FUNCT_ANNOT -t 24 -b BIN_REFINEMENT/metawrap_50_10_bins

vim annotate_bins.sh
```

# fraggenscan

```{bash}
conda install -c bioconda fraggenescan

#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/FRAGGENE
#PBS -j oe
#PBS -l nodes=node31:ppn=24,walltime=999:00:00,mem=250gb
#PBS -N FRAGGENE
#PBS -q batch
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd /home/xiongyi/YYDS/
run_FragGeneScan.pl \
  -genome=ASSEMBLY/final.contigs.fa \
  -out=FRAGGENE/YYfrag \
  -complete=0 \
  -train=complete \
  -thread=24
  
vim FRAGGENE.sh
```

# eggnog

[eggnog-mapper](https://github.com/eggnogdb/eggnog-mapper)

```{bash}
conda activate soil
conda install -c bioconda eggnog-mapper cd-hit
cd /home/xiongyi/YYDS/EGG
nohup download_eggnog_data.py \
  --data_dir /home/xiongyi/database/EGGNOG \
  -y \
  > eggDB.log.txt 2>&1 & 

#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/EGG
#PBS -j oe
#PBS -l nodes=node31:ppn=24,walltime=999:00:00,mem=250gb
#PBS -N CDHIT
#PBS -q batch
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd /home/xiongyi/YYDS/
cd-hit \
  -i FRAGGENE/YYfrag.faa \
  -o EGG/YYfrag95.faa \
  -d 0 \
  -c 0.95 \
  -aS 0.9 \
  -T 24 \
  -M 256000
  
vim CDHIT.sh

#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/EGG
#PBS -j oe
#PBS -l nodes=node31:ppn=24,walltime=999:00:00,mem=250gb
#PBS -N EGGNOG
#PBS -q batch
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd /home/xiongyi/YYDS/
emapper.py \
  --data_dir /home/xiongyi/database/EGGNOG \
  -i EGG/YYfrag95.faa \
  -o EGG/YYfrag95.annotation \
  --pident 80 \
  --evalue 1e-5 \
  --cpu 24 

vim EGGNOG.sh  
```

# kofam

```{bash}
conda install -c bioconda kofamscan
nohup wget https://www.genome.jp/ftp/db/kofam/profiles.tar.gz \
  > profiles.log.txt 2>&1 &
  
#!/bin/sh -login
#PBS -o /home/xiongyi/YYDS/EGG
#PBS -j oe
#PBS -l nodes=1:ppn=24,walltime=999:00:00,mem=120gb
#PBS -N KOFAM
#PBS -q cpu
source ~/miniconda3/etc/profile.d/conda.sh
conda activate soil
cd /home/xiongyi/YYDS/
exec_annotation \
  -f detail-tsv \
  -E 1e-5 \
  --profile /home/xiongyi/database/kofam/profiles \
  --ko-list /home/xiongyi/database/kofam/ko_list \
  --cpu 24 \
  -o EGG/YYfrag95.kegg.txt \
  EGG/YYfrag95.faa
  
vim KOFAM.sh
```

